mautic/mautic
# Repeatable campaign loop fix

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This bug has existed since 2018 when repeatable campaigns were first introduced. Repeatable campaigns are only processing 300 contacts (the batch size) at a time because the first query for applicable contacts takes into account that the campaign is repeatable (before the while loop). But the query for the next batch of contacts does not which changes the query and returns 0 contacts exiting the loop (inside the while loop).

#### Steps to test this PR:
1. Create a segment A with more than 300 contacts.
2. Create a segment B with no contacts.
3. Create a repeatable campaign with any action, possibly to add 1 point and the source is segment A.
4. Run `bin/console mautic:campaigns:rebuild` and check that all contacts from segment A are in the campaign.
5. Update the campaign source segment to segment B.
6. Run `bin/console mautic:campaigns:rebuild` and check that all contacts were removed from the campaign.
7. Update the campaign source segment back to segment A. Now the "repeatable" feature must kick in because the contacts were already in the campaign.
8. Run `bin/console mautic:campaigns:rebuild` and notice that the command progress bar goes through all the contacts, but **check how many contacts are there in the campaign actually. It is just 300, not all of them**.

##### Expected Result
The campaign should get its members within one execution of the `bin/console mautic:campaigns:rebuild` command.

##### Actual Results
The campaign adds only 300 contacts per command execution.

#### Other areas of Mautic that may be affected by the change:
1. Just campaign membership building

[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. The improved unit tests now check that the method `getCampaignContactsBySegments()` will always consider the 3rd argument in the while loop based on the campaign repeatability configuration.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10766"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
